### ISSUE TITLE
age: error: failed to wrap key for recipient #0: test plugin: write |1: broken pipe

### ISSUE THREAD

**Issue by someplaceguy** [FiloSottile/age#517]
## Environment

* OS: NixOS 23.05
* age version: 1.1.1

## What were you trying to do

Trying to run tests as part of the package's build process.

## What happened

A test failed.

```
building
Building subPackage .
Building subPackage ./agessh
Building subPackage ./armor
Building subPackage ./cmd/age
Building subPackage ./cmd/age-keygen
Building subPackage ./internal/bech32
Building subPackage ./internal/format
Building subPackage ./internal/plugin
Building subPackage ./internal/stream
running tests
ok      filippo.io/age  0.736s
ok      filippo.io/age/agessh   0.272s
ok      filippo.io/age/armor    0.016s
--- FAIL: TestScript (0.00s)
    --- FAIL: TestScript/plugin (0.06s)
        testscript.go:481: 
            # encrypt and decrypt a file with a test plugin (0.059s)
            > age -r age1test10qdmzv9q -o test.age input
            [stderr]
            age: error: failed to wrap key for recipient #0: test plugin: write |1: broken pipe
            age: report unexpected or unhelpful errors at https://filippo.io/age/report
            [exit status 1]
            FAIL: testdata/plugin.txt:2: unexpected command failure
            
FAIL
FAIL    filippo.io/age/cmd/age  1.008s
FAIL
```

The build machine may have been overloaded at the time the tests were running (e.g. load average up to ~150 on a 16-core CPU), but it should not have caused an error AFAIK.

**Comment by ctheune** [comment]
This happens repeatably (without any load issue) for me on Darwin. NixOS is on unstable.

```
Darwin chazzzzz 22.6.0 Darwin Kernel Version 22.6.0: Wed Jul  5 22:22:52 PDT 2023; root:xnu-8796.141.3~6/RELEASE_ARM64_T8103 arm64
```

Here's the full build output.

```
unpacking sources
unpacking source archive /nix/store/fxs7r50g1c64gbvnz76pc4g0kcclapaj-source
source root is source
patching sources
applying patch /nix/store/16jna1lazm41ff0rdx6dsw1h8fpwgzm2-0001-revert-temporarily-disabled-testscript-tests.patch
patching file cmd/age/age_test.go
patching file go.mod
patching file go.sum
applying patch /nix/store/6zkzhrsr4scj1dxr5igd6c513x8i3a49-0002-depend-on-c2sp_cctv_age__TestVectors.patch
patching file go.mod
patching file go.sum
patching file testkit_test.go
configuring
building
Building subPackage .
Building subPackage ./agessh
Building subPackage ./armor
Building subPackage ./cmd/age
Building subPackage ./cmd/age-keygen
Building subPackage ./internal/bech32
Building subPackage ./internal/format
Building subPackage ./internal/plugin
Building subPackage ./internal/stream
running tests
ok      filippo.io/age  1.100s
ok      filippo.io/age/agessh   0.805s
ok      filippo.io/age/armor    0.490s
--- FAIL: TestScript (0.00s)
    --- FAIL: TestScript/plugin (0.08s)
        testscript.go:481:
            # encrypt and decrypt a file with a test plugin (0.065s)
            > age -r age1test10qdmzv9q -o test.age input
            [stderr]
            age: error: failed to wrap key for recipient #0: test plugin: write |1: broken pipe
            age: report unexpected or unhelpful errors at https://filippo.io/age/report
            [exit status 1]
            FAIL: testdata/plugin.txt:2: unexpected command failure

FAIL
FAIL    filippo.io/age/cmd/age  2.166s
FAIL
```

I'm currently trying to figure out when this broke.

**Comment by ctheune** [comment]
@someplaceguy This /could/ also be a NixOS issue. Did you happen to already open a bug there?

**Comment by ctheune** [comment]
I just double-checked with the NixOS repo and our package definition hasn't changed in 8 months and has been on 1.1.1 since then: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/security/age/default.nix

**Comment by ctheune** [comment]
It does look a bit like /something/ changed on our side:
https://hydra.nixos.org/log/3p7vyypyrsbys1r4wa8zv2h9ma3cjhhd-age-1.1.1.drv

**Comment by someplaceguy** [comment]
@ctheune No, I run into these kind of non-deterministic test failures quite frequently (a few per week) due to the aforementioned CPU load issue. Apparently high CPU loads tend to be very good at finding bugs in software or even tests themselves.

All of these issues that I've ran into have been in the upstream software, never in NixOS.

But almost all of them are either test failures due to timeouts or race conditions in build systems (especially Makefiles), which I don't even bother open bugs for due to these being so common (I just disable tests or disable parallel building).

I've chosen to file this bug because it didn't seem to be either of these issues, but from superficially looking at the error, I would be surprised if the problem was in NixOS.

**Comment by ctheune** [comment]
Yeah, could be. The issue seems to trigger upstream as well and it has been happening for me since yesterday and it happens repeatedly.

Here's the build that broke upstream. https://hydra.nixos.org/build/229581170

The changes that triggered the build are here:
https://hydra.nixos.org/api/scmdiff?branch=&rev2=9462344318b376e157c94fa60c20a25b913b2381&type=git&rev1=05f42aef4e9502a0d0ad79c9debd11d98565cfa7&uri=https%3A%2F%2Fgithub.com%2FNixOS%2Fnixpkgs.git

I checked the changes but none of the commit message seem suspicuous.

I'm going to disable the tests and see whether it actually works.

BTW: the aarch-build didn't break and the pure linux builds  and I'm not quite sure whether I'm running my nix environment on MacOS with x86_64 and thus emulation ... o_O

**Comment by ctheune** [comment]
So, it seems I ended up with a working version of age when disabling the tests, but the tests keep failing reliably for me.

Maybe pinging @bdd and @zowoq who touched it last could help.

**Comment by FiloSottile** [comment]
Hi! First of all, tyhank you for your work packaging age, and for reporting things upstream.

What that test does (under a layer of abstraction in testscript) is make a copy of the test binary into a temporary directory that's added to $PATH, execute the binary with different flags that make it behave as a plugin, and communicate with it over stdin/stdout. Does any of that sound like something that would clash with the NixOS env?

**Comment by someplaceguy** [comment]
> Does any of that sound like something that would clash with the NixOS env?

Theoretically it could, as NixOS runs package builds and tests under a sandbox, but according to the information you provided it seems unlikely to me (although it's hard to tell).

>  execute the binary with different flags that make it behave as a plugin, and communicate with it over stdin/stdout. 

Again, I know almost nothing about age, but looking at this error message:

```
age: error: failed to wrap key for recipient #0: test plugin: write |1: broken pipe
```

... at first sight, if I'm interpreting the error correctly, it kinda appears that the test plugin exited prematurely because it was trying to write to its `stdout` (hence the `write` mention) but `age` must have closed its end of the pipe (hence the `broken pipe` error), which is why the `write` didn't succeed but instead aborted the execution of the test plugin.

It'd be interesting to figure out for what reasons `age` could be closing its end of the pipe, if that's the case.

**Comment by someplaceguy** [comment]
Depending on the code, another possibility is that the `write` and `broken pipe` errors really are coming from `age`, which would instead mean that it's the `test plugin` that is exiting prematurely (or at least, closing its end of the pipe) for some reason.

**Comment by bdd** [comment]
I cannot reproduce this on {x86_64, aarch64}-linux or aarch64-darwin. Neither from the trunk, nor from nixpkgs-23.05(-darwin) branches.

Looking at Hydra, this seems to fail more frequently on [x86_64-darwin](https://hydra.nixos.org/job/nixpkgs/trunk/age.x86_64-darwin/all), rarely for [aarch64-darwin](https://hydra.nixos.org/job/nixpkgs/trunk/age.aarch64-darwin/all), and seemingly never for Linux ([x86_64](https://hydra.nixos.org/job/nixpkgs/trunk/age.x86_64-linux/all), [aarch64](https://hydra.nixos.org/job/nixpkgs/trunk/age.aarch64-linux/all)).

Sounds like @ctheune can repro on their aarch64-darwin box.
I changed the derivation a bit (only running the suspect test) and patched age test suite to set `AGEDEBUG=plugin` envvar for the testscript runnner, so maybe we can catch how it fails.  Error in Hydra builds is bubbling up from `age/age.go:238` when it's  trying to wrap a file key for recipient, talking to the plugin. This happens over a pipe where plugin's input is stdin, and output is stdout.  In the test case, this plugin is simply implemented to return the key it received, without wrapping (not that it matters); and this happens over a pipe. Maybe for some reason age-test-plugin process used in the tests is dying. I'm hoping to see a bit more with the patch.

Can you please try:
```
nix build --print-build-logs --no-link github:bdd/nixpkgs/age-test-plugin#age
```
...and paste the output of the failure?

If it doesn't fail in the first case, you can try rebuilding by adding `--rebuild` to the nix build command above. 

@someplaceguy Can you try the same? I'm even more interested in seeing how this fails under Linux (as you said you're running NixOS 23.05).

**Comment by someplaceguy** [comment]
@bdd I tried both with and without your commit but I can't reproduce this issue anymore.

**Comment by squirmy** [comment]
One extra data point here. It works fine for me on aarch64-darwin. It does not work for me inside GitHub actions on x86_64-darwin.

> So, it seems I ended up with a working version of age when disabling the tests, but the tests keep failing reliably for me.

@ctheune would you mind sharing your solution for disabling the tests?

I've tried:

```nix
ageWithoutTests = pkgs.age.overrideAttrs (_: {
  doCheck = false;
});
```

But this doesn't seem to work around the issue for me.

**Comment by ctheune** [comment]
Sure. I'm using

```
    (age.overrideAttrs (finalAttrs: previousAttrs: {
      doCheck = false; # https://github.com/FiloSottile/age/issues/517
    }))
```
